contracts/evm/src/eth/utils/SparseMerkleTree.sol

```diff
@@ -0,0 +1,125 @@
+     */
+    uint8 internal constant LEAF_NODE_PREFIX = 0;
+
+    /**
+     * @notice Inner node prefix for hashing
+     */
+    uint8 internal constant INNER_NODE_PREFIX = 1;
+
+    /**
+     * @notice Verifies an SMT (non-)inclusion proof, checking that the
+     * resulting root is the same as the current root
+     * @param root Current SMT root
+     * @param proof SMT proof
+     * @return Whether the (non-)inclusion proof is valid
+     */
+    function verifyProof(bytes32 root, Proof calldata proof) internal pure returns (bool) {
+        require(proof.sideNodes.length <= MAX_DEPTH && proof.numSideNodes <= MAX_DEPTH, "Side nodes exceed depth");
+
+        bytes32 path = keccak256(abi.encodePacked(proof.key));
+        bytes32[3] memory hashBuffer;
+
+        bytes32 currentHash;
+
+        if (proof.value == DEFAULT_VALUE) {
+            if (proof.nonMembershipLeafPath == bytes32(0)) {
+                currentHash = DEFAULT_VALUE;
+            } else {
+                require(proof.nonMembershipLeafPath != path, "nonMembershipLeaf not unrelated");
+
+                currentHash =
+                    _hashNode(hashBuffer, LEAF_NODE_PREFIX, proof.nonMembershipLeafPath, proof.nonMembershipLeafValue);
+            }
+        } else {
+            currentHash = _hashNode(hashBuffer, LEAF_NODE_PREFIX, path, proof.value);
+        }
+
+        uint256 sideNodeIndex = 0;
+        for (uint256 i = 0; i < proof.numSideNodes; i++) {
+            uint256 bit = proof.numSideNodes - i - 1;
+            bytes32 sideNode = ((proof.bitMask & (1 << i)) != 0) ? DEFAULT_VALUE : proof.sideNodes[sideNodeIndex++];
+
+            if ((uint256(path) >> (255 - bit)) & 1 == 0) {
+                currentHash = _hashNode(hashBuffer, INNER_NODE_PREFIX, currentHash, sideNode);
+            } else {
+                currentHash = _hashNode(hashBuffer, INNER_NODE_PREFIX, sideNode, currentHash);
+            }
+        }
+
+        return root == currentHash;
+    }
+
+    /**
+     * @dev Hashing function for SMT nodes, which includes a prefix and uses a
+     * hash buffer previously allocated to avoiding unnecessary reallocation
+     * through `abi.encodePacked` for serializing arguments.
+     * The necessary allocation is 1 byte (prefix) + 32 bytes * 2 (left and
+     * right arguments). Following Solidity's 32-byte memory alignment, this
+     * requires a 3 slot allocation (96 bytes).
+     * Ideally this could also use free memory (in terms of free memory
+     * pointer), but memory may not be properly cleaned afterwards and it could
+     * be (wrongly) considered as clean memory when actually being allocated
+     * @param hashBuffer Memory area to be used for serialization
+     * @param prefix Node prefix for hashing
+     * @param left Left value
+     * @param right Right value
+     * @return result Result of `keccak256(prefix || left || right)`
+     */
+    function _hashNode(bytes32[3] memory hashBuffer, uint8 prefix, bytes32 left, bytes32 right)
```

Comment:
This is already the case - this just passes a pointer to a previously reserved memory slot. We can't really use the scratch area in this case though, it's only 64 bytes.

Reply:
I could see that its previously allocated memory) Was wondering if there anything similiar to static variables in solidity not to argument around? 

Are variables declared inside contract(not functions) only "storage" variables rr it can be "memory"?

Reply:
Got what you mean. We can't do it in Solidity afaik - contract variables are storage-based, and there's no way to make a static variable so we didn't need to pass the pointer around.